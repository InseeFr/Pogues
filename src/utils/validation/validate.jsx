// @TODO: Refactor to avoid similar validation functions
import get from 'lodash.get';
import { SubmissionError } from 'redux-form';

import {
  DATATYPE_NAME,
  QUESTION_TYPE_ENUM,
  TABS_PATHS,
} from '../../constants/pogues-constants';
import { Component } from '../../model';
import {
  calculatedVariableRules,
  collectedVariableRules,
  controlRules,
  declarationRules,
  durationRulesPTnHnM,
  durationRulesPnYnM,
  externalVariableRules,
  filterRules,
  loopRules,
  questionRules,
  questionnaireRules,
  redirectionRules,
  roundaboutRules,
  sequenceRules,
  tableListMeasuresRules,
} from './validation-rules';
import { getErrorsObject, validate } from './validation-utils';

export function validateQuestionnaireForm(values, setErrors) {
  const errors = validate(values, questionnaireRules);

  if (errors.length > 0) {
    setErrors(errors);
    throw new SubmissionError(getErrorsObject(errors));
  }
}

const { SIMPLE } = QUESTION_TYPE_ENUM;
const { DURATION } = DATATYPE_NAME;
const { RESPONSE_FORMAT } = TABS_PATHS;

export function validateQuestionForm(values, setErrors, codesListsStore) {
  let errors;
  const durationFormat = get(
    values,
    `${RESPONSE_FORMAT}.${SIMPLE}.${DURATION}.format`,
  );
  if (durationFormat === 'PTnHnM') {
    errors = validate(
      values,
      { ...questionRules, ...durationRulesPTnHnM },
      { codesListsStore },
    );
  } else if (durationFormat === 'PnYnM') {
    errors = validate(
      values,
      { ...questionRules, ...durationRulesPnYnM },
      { codesListsStore },
    );
  } else {
    errors = validate(values, questionRules, { codesListsStore });
  }
  if (errors.length > 0) {
    setErrors(errors);
    throw new SubmissionError(getErrorsObject(errors));
  }
}

export function validateSequenceForm(values, setErrors) {
  const errors = validate(values, sequenceRules);

  if (errors.length > 0) {
    setErrors(errors);
    throw new SubmissionError(getErrorsObject(errors));
  }
}

export function validateLoopForm(values, setErrors) {
  const errors = validate(values, loopRules);
  if (errors.length > 0) {
    setErrors(errors);
    throw new SubmissionError(getErrorsObject(errors));
  }
}

export function validateFilterForm(values, setErrors) {
  const errors = validate(values, filterRules);
  if (errors.length > 0) {
    setErrors(errors);
    throw new SubmissionError(getErrorsObject(errors));
  }
}

export function validateRoundaboutForm(values, setErrors) {
  const errors = validate(values, roundaboutRules);
  if (errors.length > 0) {
    setErrors(errors);
    throw new SubmissionError(getErrorsObject(errors));
  }
}

// Subforms validation

export function validateControlForm(values, addErrors) {
  const errors = validate(values, controlRules);

  // SubmissionError can't be used in subforms validations
  if (errors.length > 0) {
    addErrors(errors);
    return false;
  }
  return true;
}

export function validateDeclarationForm(values, addErrors) {
  const errors = validate(values, declarationRules);

  // SubmissionError can't be used in subforms validations
  if (errors.length > 0) {
    addErrors(errors);
    return false;
  }
  return true;
}

export function validateRedirectionForm(values, addErrors, stores) {
  const errors = validate(values, redirectionRules, stores);

  // SubmissionError can't be used in subforms validations
  if (errors.length > 0) {
    addErrors(errors);
    return false;
  }
  return true;
}

export function validateCalculatedVariableForm(values, addErrors, state) {
  const errors = validate(values, calculatedVariableRules, {}, state);

  // SubmissionError can't be used in subforms validations
  if (errors.length > 0) {
    addErrors(errors);
    return false;
  }
  return true;
}

export function validateExternalVariableForm(values, addErrors, state) {
  const errors = validate(values, externalVariableRules, {}, state);
  // SubmissionError can't be used in subforms validations
  if (errors.length > 0) {
    addErrors(errors);
    return false;
  }
  return true;
}

export function validateCollectedVariableForm(values, addErrors, state) {
  const errors = validate(values, collectedVariableRules, {}, state);

  // SubmissionError can't be used in subforms validations
  if (errors.length > 0) {
    addErrors(errors);
    return false;
  }
  return true;
}

export function validateTableListMeasuresForm(values, addErrors) {
  const normalizedValues = Component().getNormalizedValues(values);
  const errors = validate(normalizedValues, tableListMeasuresRules);
  // SubmissionError can't be used in subforms validations
  if (errors.length > 0) {
    addErrors(errors);
    return false;
  }
  return true;
}
