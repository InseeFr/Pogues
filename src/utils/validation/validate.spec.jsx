import { describe, expect, test } from 'vitest';

import {
  DATATYPE_NAME,
  QUESTION_TYPE_ENUM,
  TABS_PATHS,
} from '../../constants/pogues-constants';
import { validateQuestionForm } from './validate';

const { SIMPLE } = QUESTION_TYPE_ENUM;
const { DURATION } = DATATYPE_NAME;
const { RESPONSE_FORMAT } = TABS_PATHS;

const fakeValuePTnHnM = {
  [RESPONSE_FORMAT]: {
    [SIMPLE]: {
      [DURATION]: {
        format: 'PTnHnM',
        maminutes: '',
        miminutes: '',
        mahours: '',
        mihours: '',
      },
    },
  },
};

const fakeValuePnYnM = {
  [RESPONSE_FORMAT]: {
    [SIMPLE]: {
      [DURATION]: {
        format: 'PnYnM',
        mamonths: '',
        mimonths: '',
        mayears: '',
        miyears: '',
      },
    },
  },
};

describe('Special validation for duration', () => {
  test('should validate have right rules for duration (PTnHnM) question (required)', () => {
    let validationErrors;
    const setErrors = (errors) => {
      validationErrors = errors;
    };

    expect(() =>
      validateQuestionForm(fakeValuePTnHnM, setErrors, {}),
    ).toThrowError(/Submit Validation Failed/);

    expect(validationErrors).toMatchSnapshot();
  });

  test('should validate have right rules for duration (PnYnM) question (required)', () => {
    let validationErrors;
    const setErrors = (errors) => {
      validationErrors = errors;
    };

    expect(() =>
      validateQuestionForm(fakeValuePnYnM, setErrors, {}),
    ).toThrowError(/Submit Validation Failed/);

    expect(validationErrors).toMatchSnapshot();
  });
});
