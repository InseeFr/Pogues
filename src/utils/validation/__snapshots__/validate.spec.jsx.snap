// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Special validation for duration > should validate have right rules for duration (PTnHnM) question (required) 1`] = `
[
  {
    "errors": [
      "Mandatory field",
    ],
    "path": "responseFormat.SIMPLE.DURATION.maminutes",
  },
  {
    "errors": [
      "Must be at least 0",
    ],
    "path": "responseFormat.SIMPLE.DURATION.maminutes",
  },
  {
    "errors": [
      "Must be at most 59",
    ],
    "path": "responseFormat.SIMPLE.DURATION.maminutes",
  },
  {
    "errors": [
      "Mandatory field",
    ],
    "path": "responseFormat.SIMPLE.DURATION.miminutes",
  },
  {
    "errors": [
      "Must be at least 0",
    ],
    "path": "responseFormat.SIMPLE.DURATION.miminutes",
  },
  {
    "errors": [
      "Must be at most 59",
    ],
    "path": "responseFormat.SIMPLE.DURATION.miminutes",
  },
  {
    "errors": [
      "Mandatory field",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mahours",
  },
  {
    "errors": [
      "Mandatory field",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mihours",
  },
]
`;

exports[`Special validation for duration > should validate have right rules for duration (PnYnM) question (required) 1`] = `
[
  {
    "errors": [
      "Mandatory field",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mamonths",
  },
  {
    "errors": [
      "Must be at least 0",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mamonths",
  },
  {
    "errors": [
      "Must be at most 11",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mamonths",
  },
  {
    "errors": [
      "Mandatory field",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mimonths",
  },
  {
    "errors": [
      "Must be at least 0",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mimonths",
  },
  {
    "errors": [
      "Must be at most 11",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mimonths",
  },
  {
    "errors": [
      "Mandatory field",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mayears",
  },
  {
    "errors": [
      "Must be at least 0",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mayears",
  },
  {
    "errors": [
      "Must be at most 11",
    ],
    "path": "responseFormat.SIMPLE.DURATION.mayears",
  },
  {
    "errors": [
      "Mandatory field",
    ],
    "path": "responseFormat.SIMPLE.DURATION.miyears",
  },
]
`;
